### DISCUSSION TITLE
[Technical Questions] How to write tests for codes executed only after getting HealthKit Data

### DISCUSSION THREAD

**Discussion by YurenSUN** [CS342/.github#68] [Technical Questions] [answered]
### In what area do you have a technical challenge?

UI Testing related to HealthKit

### Description

Hi, when I am developing the visualization for HealthKit data, I find it hard to test my codes to parse and plot the queried HealthKit data. For example, in the Codecov report ([1](https://app.codecov.io/gh/CS342/2024-PICS/pull/31/blob/PICS/HealthVisulization/HKVisualizationItem.swift) and [2](https://app.codecov.io/gh/CS342/2024-PICS/pull/31/blob/PICS/HealthVisulization/HKVisualization.swift) for this [PR](https://github.com/CS342/2024-PICS/pull/31)), we can see that the codes marked red (missing coverages) are all related to either reading the HealthKit query results, or plotting the HealthKit data.

I wrote a simple UI [test](https://github.com/CS342/2024-PICS/blob/main/PICSUITests/HealthVisualizationTests.swift) to check whether the dashboard is loaded as expected but missing coverage for those codes as I do not know whether there are HealthKit data available in PR and Codecov testing.

Is there a desired way to test such codes (e.g. maybe manually input health data and check, or initialize those views directly with faked health kit data instead of starting the whole application in the UI tests, or could this be done with unit tests), and/or are there any examples for testing such codes?

### Reproduction

The issues could be viewed from the Codecov report ([1](https://app.codecov.io/gh/CS342/2024-PICS/pull/31/blob/PICS/HealthVisulization/HKVisualizationItem.swift), and [2](https://app.codecov.io/gh/CS342/2024-PICS/pull/31/blob/PICS/HealthVisulization/HKVisualization.swift)). It is clearer in the second link that the functions marked red are just functions processing the fetched HealthKit data.

### Expected behavior

If possible, I hope to have the codes tested possibly with faked HealthKit data. Thank you! 😄 

### Code of Conduct

- [X] I agree to follow this course's Code of Conduct and Contributing Guidelines

**Comment by PSchmiedmayer** [comment, marked as answer]
Hi @YurenSUN, a great question! 

You can manually create a lot of the HealthKit data types and manually call the functions you want to call within a unit test.
E.g., you can create a simple HKSample like this: https://developer.apple.com/documentation/healthkit/hkquantitysample/1615016-init.

For those that you cannot create using code or when you need to do a full integration test: We have a dedicated packe for this: https://github.com/StanfordBDHG/XCTHealthKit. It allows you to run UI tested that add elements from the health app.

**Reply by YurenSUN** [reply]
Sounds great! Thank you @PSchmiedmayer!
